### DISCUSSION TITLE
What's the best way to avoid CFL-related instabilities in the early simulation steps of LES

### DISCUSSION THREAD

**Discussion by tomchor** [CliMA/Oceananigans.jl#1361] [Computational science!] [answered]
I'm running a LES of the upper ocean with a Smagorinsky closure. In the beginning of the simulations the Courant number varies a lot (probably because of the small-scale noise) so I have to set the CFL to something really small to get passed that. (Something like 0.02!) The problem is that eventually the Courant number stabilizes and I get stuck with a simulation that takes _much_ smaller steps than it could, and therefore doesn't use computational resources efficiently.

In other codes that I know this is generally counter-acted by starting the first couple hundred steps of an LES with a constant eddy diffusivity, and only then turning on the LES closure. Is there an easy way to do that with Oceananigans? Sorry if I missed something, but I searched the docs and couldn't find anything.

Ideally I'd wanna avoid starting a very brief simulation with a constant viscosity and then pickup that simulation with another script to run it "for real". Alternatively having a time-dependent CFL limit might work, but that seems unnecessarily convoluted...

Thoughts?

**Comment by ali-ramadhan** [comment, marked as answer]
I've used a time-dependent CFL in the past with some good results. It's easy to implement in the simulation progress function but a bit awkward. This is how I did it for the stratified Couette flow validation experiment (the script needs some real TLC): https://github.com/CliMA/Oceananigans.jl/blob/cadfdaee11241b887bdaef7343eaede047248b85/validation/stratified_couette_flow/stratified_couette_flow.jl#L237-L244

Switching model properties is not allowed since that would modify the model's Julia type. You also might not want to create two models as that would allocate twice as much memory.

Might not be too hard to write a function like
```julia
new_model = switch_out_model(old_model, closure=new_closure, ...)
```
that constructs and returns a new model using all the arrays from the old model to avoid allocating any new memory.

It would perform a somewhat similar job to the checkpointer `set!` function: https://github.com/CliMA/Oceananigans.jl/blob/master/src/OutputWriters/checkpointer.jl#L236

**Reply by tomchor** [reply]
That's interesting, thanks for sharing this script! Apparently doing a time-dependent CFL is easier than I thought. (And yes, creating two models crossed my mind but indeed I cannot spare the extra memory required.) Were you trying to solve the same problem that I'm facing?  I imagine so, since this is a common LES issue.

Under the same context as https://github.com/CliMA/Oceananigans.jl/pull/999, do you think it's possible to instantiate a model such as the pseudocode below?

```julia
model = IncompressibleModel(architecture=arch, grid=grid, 
                                             closure= time < t0 ? closure_1 : closure_2)
```

I think if this was possible it would be a pretty good solution with possibly other uses.

**Reply by ali-ramadhan** [reply]
Yeah it might have been a similar issue: transition to turbulence was causing the model to blow up so I ramped up CFL linearly to some max.

For other simulations I've also done stuff like increasing the wizard's `max_Δt` after some safe time past the transition:

```julia
wizard = TimeStepWizard(..., max_Δt=some_small_value)  # initially

function print_progress(simulation)
    ...
    model.clock.time > t_transition && (wizard.max_Δt = some_larger_value)
    ...
end
```

I think supporting something like `closure = time < t0 ? closure_1 : closure_2` might be more trouble than it's worth but @glwagner might be able to comment better.

**Reply by tomchor** [reply]
> I think supporting something like closure = time < t0 ? closure_1 : closure_2 might be more trouble than it's worth but @glwagner might be able to comment better.

Yeah, I'm not sure what are the implications of that. 

I'm pretty sure this issue of starting a LES from rest with only noise is very common though. In all the LES codes I've ever used this problem happened and it was solved by starting with a simpler closure after a few iterations. So I think it'd be really useful for users in general if Oceananigans permitted a better solution than limiting the CFL in the beginning. What do you think?

**Reply by ali-ramadhan** [reply]
Right if it's an important use case then it shouldn't be too hard to support. It would certainly simplify some of the simulations I run.

I guess I'm not sure what the backend logic would look like. Making `model.closure` a function might complicate things for the Julia compiler?

I'd advocate for developing a function like `switch_out_model` for changing model properties as it can then be used for other properties beyond just the `closure` and it might be easier to support more complex logic, i.e. change closure based on not just the time.

Then the user interface would look like `closure = time < t0 ? closure_1 : closure_2` but behind the scenes `switch_out_model(old_model, closure = closure2)`  would get used somewhere.

This could all be implemented rather cleanly using simulation callbacks but that would require #1138 to get resolved.

**Reply by tomchor** [reply]
I really don't know enough about Oceananigans/Julia to be of much help on the details of the implementation. I do believe that users (like me) would benefit a lot from some relatively simple solution since this should be a common problem.

I'm assuming #1138 will take a while to implement, no? In that context, is there any other workarounds that you recommend? I'm trying to limit the CFL like you showed me. It works (and I appreciate it!) but it's still wasting a lot of computational power since I have to keep the CFL low for many thousands of steps, instead of doing a constant diffusivity for a couple hundred time steps.

Thanks for the help!

**Reply by glwagner** [reply]
For problems like these I often start with a very small time-step and adjust `max_change` so that it ramps up slowly to my target `CFL`.

#1138 doesn't provide a _substantive_ change to the API, it just makes things easier / cleaner to implement.

I'm not 100% sure what's being asked for. The issue refers to CFL-related instabilities, but I think what is being asked for is a turbulence closure that reverts from a constant diffusivity to a nonlinear LES diffusivity at some specified `transition_time`?

One way to implement such a model right now is to build two models, one with a constant diffusivity, and one with an LES closure. Then users can `run!` the constant diffusivity model until `transition_time`, copy over model data (just `set!(new_model, fields(old_model)...)` and also setting `new_model.clock.time` and `new_model.clock.iteration`) and then continuing the run. If the time and iteration are set then the same output writers can be used for the new simulation, which would produce a continuous time series in output. I think. Things like building a `NamedTuple` of `kwargs` will help make the script short-ish and easy to configure.

The above design does waste some memory. To avoid this, users can build the new model using something like

```julia
new_model = IncompressibleModel(old_model_kwargs..., closure=SmagorinskyLilly(), velocities=old_model.velocities, tracers=old_model.tracers, pressures=old_model.pressures, clock=old_model.clock)
```

I personally think this is a nice solution since it explicitly reflects a sensible physical / modeling interpretation of this approach: we have two different physical models that are "stitched" together at a certain `transition_time`.

Defining a new closure is also possible. However, this would require more work than the above options because the appropriate closure functions would/ might have to be defined. Another difficulty is that the syntax of `calculate_diffusivities!` assumes the turbulent diffusivity is independent of time. All of these can be solved "easily" (no magic needed, just more code), but I think it would require more work than above.

A callback that manually adjusts the diffusivities would only work for Adams-Bashforth time-stepping, because the diffusivity is not updated during the substages of the RK3 time-stepping. It could be worth incorporating such a feature into #1138 (providing a special list of callbacks that is executed before every explicit substage?)

**Reply by tomchor** [reply]
@glwagner thanks for the suggestion, but I get an out-of-memory error doing this, so I think it still adds a memory burden (or I'm dong something wrong)

> I'm not 100% sure what's being asked for. The issue refers to CFL-related instabilities, but I think what is being asked for is a turbulence closure that reverts from a constant diffusivity to a nonlinear LES diffusivity at some specified transition_time?

I'm just trying to come up with a good way to avoid the model blowing up in the initial steps while still efficient with the time-stepping. So far I guess we have two solutions:

- Limiting the CFL in the beginning. It works, it doesn't add a memory burden, but it's inefficient in terms of the time-stepping
- Creating a preliminary model with a constant diffusivity and then create another model based on it, but with a dynamic closure works, but apparently adds a memory burden
- Swapping diffusivity closures (from constant to dynamic) in the _same model_ works, since other models do that, but apparently can't be done easily in Oceananigans for now.

@ali-ramadhan how hard would be to write my own `switch_out_model` function until #1138 is resolved? I wouldn't really know where to start.

**Reply by glwagner** [reply]
The pressure solver has some memory associated with it, so you can try `pressure_solver = old_model.pressure_solver`.

If that fails you can invoke the baseline `IncompressibleModel` constructor directly.

There's no effective difference between "swapping" closures in two different models, or just building a new model with a different closure.

**Reply by glwagner** [reply]
The full list of kwargs is here:

https://github.com/CliMA/Oceananigans.jl/blob/86a54b7b7bc591a1706946330323441fb4db5c1f/src/Models/IncompressibleModels/incompressible_model.jl#L82-L102

**Reply by glwagner** [reply]
So maybe you need `diffusivities=old_model.diffusivities` as well.

**Reply by glwagner** [reply]
If that fails you can replicate what's used by the outer `IncompressibleModel` constructor:

https://github.com/CliMA/Oceananigans.jl/blob/86a54b7b7bc591a1706946330323441fb4db5c1f/src/Models/IncompressibleModels/incompressible_model.jl#L150-152

replacing all model properties with `old_model.***` except for `closure`, which you'll replace with `new_closure`.

What does "model" mean in the context of "other codes"? Here when we refer to a model we are referring specifically to an instance of the concretely-typed object `IncompressibleModel`. Functionally you can certainly achieve the same thing as another code by building an `IncompressibleModel` using existing structures in memory.

**Reply by glwagner** [reply]
Ah and also `timestepper=old_model.timestepper`...

**Reply by glwagner** [reply]
This actually might be a three-liner (but I'm not 100% sure without testing it):

```julia
new_model_specification = Dict(property_name => getproperty(old_model, name) for name in propertynames(old_model))

new_model_specification[:closure] = new_closure

new_model = IncompressibleModel(; new_model_specification...)
```

**Reply by glwagner** [reply]
Those three lines could be the basis for a nice convenience function design

```julia
function similar_model(old_model::IncompressibleModel; kwargs...)

    new_model_specification = Dict(property_name => getproperty(old_model, name) for name in propertynames(old_model))

    for (name, property) in kwargs
        new_model_specification[name] = property
    end
    
    return IncompressibleModel(; new_model_specification...)
end
```

(aka `switch_out_model`). Then you'd write

```julia
new_model = similar_model(old_model; closure=SmagorinskyLilly())
```

Just a PR plum waiting to be picked...

**Reply by tomchor** [reply]
I just coded something similar in parallel after reading your remark about the pressure solver. So I guess we think at least a bit alike haha. Your coding is much more elegant though, so I'm definitely gonna adapt mine :)

That said, I still haven't been able to make it work. But I'm working on it and I think this time I won't get the memory error if what you're saying is true. 

In the meantime, if this indeed works, should this be added to LESbrary? I think this solves an issue that's specific to large eddies simulations.

**Reply by glwagner** [reply]
I think `similar_model` is appropriate for the `Oceananigans.IncompressibleModels` module as a start, what do you think @ali-ramadhan ?

I think the outer constructor can be bypassed by modifying `similar_model` to something like

```julia
using OrderedDict

function similar_model(old_model::IncompressibleModel; kwargs...)

    new_model_specification = OrderedDict(property_name => getproperty(old_model, name) for name in propertynames(old_model))

    for (name, property) in kwargs
        new_model_specification[name] = property
    end
    
    return IncompressibleModel(values(new_model_specification)...)
end
```

This bypasses the kwarg-based constructor. But we need `OrderedDict` to make sure the properties are given in the right order. I think.

But if we can't get the kwarg-based constructor to work I think there's a problem with it. It's been designed with a very similar use case in mind (restoring checkpointed models).

**Reply by tomchor** [reply]
Apparently the code works, and it indeed doesn't seem to make me run out of memory :+1:. The code I ended up with is

```julia
function swap_model(old_model::IncompressibleModel; kwargs...)
    new_model_specification = Dict(property_name => getproperty(old_model, property_name)
                                   for property_name in propertynames(old_model))

    for (property_name, property) in kwargs
        new_model_specification[property_name] = property
    end

    return IncompressibleModel(; new_model_specification...)
end
```

For some reason I'm running into trouble with some stuff that works in the first model though. So apparently taking this approach makes writing diagnostics a bit trickier. For example:

```julia
julia> model.velocities.u + model.background_fields.velocities.u
BinaryOperation at (Face, Center, Center)
├── grid: RegularCartesianGrid{Float64, Periodic, Bounded, Bounded}(Nx=8, Ny=256, Nz=8)
│   └── domain: x ∈ [0.0, 312.5], y ∈ [0.0, 10000.0], z ∈ [-50.0, 0.0]
└── tree: 
    + at (Face, Center, Center) via identity
    ├── Field located at (Face, Center, Center)
    └── FunctionField located at (Face, Center, Center)
├── func: u_g
├── grid: RegularCartesianGrid{Float64, Periodic, Bounded, Bounded}(Nx=8, Ny=256, Nz=8)
├── clock: Clock(time=0 seconds, iteration=0)
└── parameters: nothing

julia> model = swap_model(model; closure=closure_LES)
IncompressibleModel{CPU, Float64}(time = 0 seconds, iteration = 0) 
├── grid: RegularCartesianGrid{Float64, Periodic, Bounded, Bounded}(Nx=8, Ny=256, Nz=8)
├── tracers: (:b,)
├── closure: SmagorinskyLilly{Float64,NamedTuple{(:b,),Tuple{Float64}},NamedTuple{(:b,),Tuple{Float64}}}
├── buoyancy: BuoyancyTracer
└── coriolis: FPlane{Float64}

julia> model.velocities.u + model.background_fields.velocities.u
ERROR: type ZeroField has no field grid
Stacktrace: ...
```

What am I missing here?

**Reply by ali-ramadhan** [reply]
I think `propertynames` returns the properties in the order defined in the `struct` which might be different from the order in the constructor.

My only gripe with `similar_model` is that it seems analagous to the `similar` function which would suggest it creates an "uninitialized" model:

```
help?> similar                                                                                                                                                                                                                
search: similar
                       
  similar(array, [element_type=eltype(array)], [dims=size(array)])
                        
  Create an uninitialized mutable array with the given element type and size, based upon the given source array.
```

**Reply by glwagner** [reply]
Order shouldn't matter if we are using kwargs though.

@ali-ramadhan fair point that `similar` (the function I was thinking of) returns an uninitialized object. Perhaps `copy_model`? `swap_model_properties` could work too.

@tomchor I think the problem is that the [constructor assumes `background_fields` kwarg / user input is "flat"](https://github.com/CliMA/Oceananigans.jl/blob/86a54b7b7bc591a1706946330323441fb4db5c1f/src/Fields/background_fields.jl#L23), but then gathers the background fields into `velocities` and `tracers` categories. (This is a good lesson that its best for user input to correspond as closely as possible to the object that's created.) So you'd need to write:

```julia
new_model = copy_model(old_model; closure=SmagorinskyLilly(),
                       background_fields=merge(old_model.background_fields.velocities, old_model.background_fields.tracers))
```

Smoothing out the user experience could be part of the PR for this feature. There will still be some bumps, especially due to the fact that `model.diffusivities` depends on `model.closure`. This would be fixed by storing the eddy diffusivities in `model.closure` rather than `model.diffusivities` (which we know how to do now since we now know how to use `adapt_structure` properly), which hopefully we'll have before too long.

**Reply by tomchor** [reply]
@glwagner Just FYI, that didn't work. Here's an example with a more simple code:

```julia
julia> new_model  = swap_model(model; closure=closure_LES,
                          background_fields=(u=old_model.background_fields.velocities.u, b=old_model.background_fields.tracers.b))
ERROR: MethodError: no method matching regularize_background_field(::Type{Face}, ::Type{Center}, ::Type{Center}, ::Oceananigans.Fields.FunctionField{Face,Center,Center,Clock{Float64},Nothing,typeof(u_g),RegularCartesianGrid{Float64,Periodic,Bounded,Bounded,OffsetArrays.OffsetArray{Float64,1,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}}, ::RegularCartesianGrid{Float64,Periodic,Bounded,Bounded,OffsetArrays.OffsetArray{Float64,1,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}, ::Clock{Float64})
Closest candidates are:
  regularize_background_field(::Any, ::Any, ::Any, ::BackgroundField{var"#s64",P} where P where var"#s64"<:Function, ::Any, ::Any) at /glade/u/home/tomasc/.julia/packages/Oceananigans/iSJKi/src/Fields/background_fields.jl:59
  regularize_background_field(::Any, ::Any, ::Any, ::Function, ::Any, ::Any) at /glade/u/home/tomasc/.julia/packages/Oceananigans/iSJKi/src/Fields/background_fields.jl:62
...
```

So we apparently need to pass the functions:

```julia
julia> new_model  = swap_model(model; closure=closure_LES,
                          background_fields=(u=old_model.background_fields.velocities.u.func, b=old_model.background_fields.tracers.b.func))
```

Although this little quirk can be dealt with inside the function I guess, so that the user doesn't need to worry about it.

> This would be fixed by storing the eddy diffusivities in model.closure rather than model.diffusivities (which we know how to do now since we now know how to use adapt_structure properly), which hopefully we'll have before too long.

I upvote that!

**Reply by glwagner** [reply]
We can hack around it, but I think a more robust and long-term improvement is to change the way that `background_fields` are parsed. Fixing that up might have additional benefits down the line.

**Reply by tomchor** [reply]
That sounds okay with me. Just for the record, this worked, and now my simulation runs much smoother. Should I do something with the code for now? I feel that until we make it work seamlessly (meaning that no special treatment of the background fields is required of the user) I shouldn't add it to Oceananigans. But I don't know if everyone agrees.

**Reply by glwagner** [reply]
I think adding

```julia
BackgroundFields(background_fields::NamedTuple{(:velocities, :tracers)}, args...) = background_fields
```

to

https://github.com/CliMA/Oceananigans.jl/blob/df9b830e574261efb1b7b0ee43b783be2381ceaf/src/Fields/background_fields.jl

will make `copy_model` work without needing special translation for `background_fields`. This makes is so that this line:

https://github.com/CliMA/Oceananigans.jl/blob/df9b830e574261efb1b7b0ee43b783be2381ceaf/src/Models/IncompressibleModels/incompressible_model.jl#L140

in the constructor for `IncompressibleModel` returns the right thing when `background_fields` is already "regularized" and organized into a named tuple with fields `velocities` and `tracers`.

There are other bumps that might arise since we not covering all cases in this discussion.

More broadly this suggests a flaw in the design of `BackgroundFields`, but fixing that deeper issue is for another time / PR.
